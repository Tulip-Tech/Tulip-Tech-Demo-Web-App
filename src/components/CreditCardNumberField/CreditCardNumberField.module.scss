.cardTypeWrapper {
  display: flex;
  justify-content: center;
  align-items: center;
  margin-right: 8px;
}
