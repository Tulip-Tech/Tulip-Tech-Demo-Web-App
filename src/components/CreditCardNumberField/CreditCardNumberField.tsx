import React, { useState } from 'react';
import Payment from 'payment';

import TextField from '../TextField/TextField';

import styles from './CreditCardNumberField.module.scss';

import { getPublicUrl } from '#src/utils/domHelpers';

type Props = {
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  onBlur?: React.FocusEventHandler<HTMLInputElement>;
  error?: string;
  value?: string;
};

const cardIssuers: { [key: string]: string } = {
  visa: getPublicUrl('/images/payments/visa.svg'),
  mastercard: getPublicUrl('/images/payments/mastercard.svg'),
  maestro: getPublicUrl('/images/payments/maestro.svg'),
  amex: getPublicUrl('/images/payments/amex.svg'),
  discover: getPublicUrl('/images/payments/discover.svg'),
  diners: getPublicUrl('/images/payments/diners.svg'),
  dinersclub: getPublicUrl('/images/payments/diners.svg'),
  unionpay: getPublicUrl('/images/payments/unionpay.svg'),
  hiper: getPublicUrl('/images/payments/hiper.svg'),
};

const CreditCardNumberField: React.FC<Props> = ({ value, error, onChange, onBlur, ...props }: Props) => {
  const [cardIssuer, setCardIssuer] = useState<JSX.Element | null>(null);

  const formatCreditCardNumber: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const issuer = Payment.fns.cardType(e.target?.value);
    const clearValue = e.target?.value.replace(/\D+/g, '');

    let nextValue;
    setCardIssuer(
      cardIssuers?.[issuer] ? (
        <div className={styles.cardTypeWrapper}>
          <img src={cardIssuers[issuer]} alt={issuer} height="25" />
        </div>
      ) : null,
    );

    switch (issuer) {
      case 'amex':
        nextValue = `${clearValue.slice(0, 4)} ${clearValue.slice(4, 10)} ${clearValue.slice(10, 15)}`;
        break;
      case 'dinersclub':
        nextValue = `${clearValue.slice(0, 4)} ${clearValue.slice(4, 10)} ${clearValue.slice(10, 14)}`;
        break;
      default:
        nextValue = `${clearValue.slice(0, 4)} ${clearValue.slice(4, 8)} ${clearValue.slice(8, 12)} ${clearValue.slice(12, 16)}`;
        break;
    }

    if (onChange) {
      e.target.value = nextValue.trim();
      onChange(e);
    }
  };

  return (
    <TextField
      value={value}
      name="cardNumber"
      label={`Card number`}
      onChange={formatCreditCardNumber}
      onBlur={onBlur}
      error={!!error}
      helperText={error ? error : null}
      pattern="\d*"
      {...props}
      type="text"
      placeholder="1234 5678 9012 3456"
      rightControl={cardIssuer}
      required
    />
  );
};

export default CreditCardNumberField;
