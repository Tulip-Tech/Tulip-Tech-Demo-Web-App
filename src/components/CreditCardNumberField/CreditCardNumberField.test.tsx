import React from 'react';
import { render } from '@testing-library/react';

import CreditCardNumberField from './CreditCardNumberField';

describe('<CreditCardNumberField>', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(<CreditCardNumberField value="" onChange={vi.fn()} />);

    expect(container).toMatchSnapshot();
  });
});
