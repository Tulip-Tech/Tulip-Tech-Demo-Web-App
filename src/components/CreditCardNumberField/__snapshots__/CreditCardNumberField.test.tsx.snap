// Vitest Snapshot v1

exports[`<CreditCardNumberField> > renders and matches snapshot 1`] = `
<div>
  <div
    class="_textField_e16c1b"
  >
    <label
      class="_label_e16c1b"
      for="text-field_1235_cardNumber"
    >
      Card number
    </label>
    <div
      class="_container_e16c1b"
    >
      <input
        class="_input_e16c1b"
        id="text-field_1235_cardNumber"
        name="cardNumber"
        pattern="\\\\d*"
        placeholder="1234 5678 9012 3456"
        required=""
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
